text/0000-run-to-completion-async-fn.md

```diff
@@ -0,0 +1,602 @@
+macro to cancel operations/`Future`s which do not make use of
+run-to-completion semantics. This would for example be async channels, timers,
+mutexes, semaphores, etc.
+
+In order to cancel run to completion operations, the `CancellationToken` must be 
+forwarded up to a point where cancellation is checked. This would e.g.
+require a library which e.g. utilizes `io_uring` to perform async IO operations
+to take a `CancellationToken` as an argument and listen for the cancellation
+request while the operation is still pending. The signature of such a method
+would e.g. be
+
+```rust
+async completion fn read_with_uring(
+    &mut self, buffer: &mut [u8], cancel_token: &CancellationToken
+) -> Result<usize, IoError>;
+```
+
+The implementation of such a method would rather be complicated - but only needs
+to occur within the runtime which provides the method. Most end user code is
+simply expected to forward the `CancellationToken`:
+
+```rust
+async completion fn read_all_with_uring(
+    reader: &mut Reader, buffer: &mut [u8], cancel_token: &CancellationToken
+) -> Result<(), IoError> {
+    let mut offset = 0;
+    while offset != buffer.len() {
+        let read = reader.read_with_uring(&buffer[offset..], cancel_token)?;
+        offset += read;
+    }
+    Ok(())
+}
+```
+
+Runtimes could also decide to forward `CancellationToken`s implicitely through
+task-local storage. Or the ability to forward a `CancellationToken` could be
+added to the [`std::task::Context`](https://doc.rust-lang.org/1.42.0/std/task/struct.Context.html)
+type in the future. Such a mechanism is however outside of the scope of this
+proposal.
+
+## Timeout handling with `async completion fn`
+
+Another common case where a `select!` like functionality is needed besides
+checking for cancellation is for timeouts. The current `select!` macro could not
+be used to time-out on `async completion fn`, since `select!` does not drive
+the aborted branches to completion.
+
+**Therefore the following example is not valid**:
+
+```rust
+async completion fn read_with_timeout(
+    reader: &mut Reader, buffer: &mut [u8], timeout: Duration
+) -> Result<usize, IoError> {
+    select! {
+        read_result = reader.read_with_uring(&buffer[offset..]) => read_result
+        _ = timer.delay_for(timeout) => Err(IoError::Timeout)
+    }
+}
+```
+
+Instead of this, a new mechanism is required which guarantees that the non
+timeout branches will be cooperatively cancelled and driven to completion after
+the timeout occured. That mechanism can be provided by runtimes in a variety of
+fashions. They could either provide a variant of the `select!` macro which is
+usable for `async completion fn`. Or they could provide a dedicated timeout
+method:
+
+```rust
+async completion fn read_with_timeout(
+    reader: &mut Reader, buffer: &mut [u8], timeout: Duration
+) -> Result<usize, IoError> {
+    runtime.with_timeout(timeout, async completion move {
```

Comment:
where does runtime here come from?

Reply:
Good catch. I will just make it `runtime::with_timeout`. Otherwise it would need to be passed